### PR TITLE
Use anyhow for error handling in pRuntime and pHost

### LINKED ISSUE #174: Improve error handling with anyhow
Now Phala uses Result<> and Error enum to handle errors. It requires tedious type conversion and wrapping, and there's no stack backtrace in any error. [Anyhow](https://docs.rs/anyhow/1.0.39/anyhow/) solves this problem.

Std problem: phost is `std`, and thus easy to migrate to anyhow. However pruntime is `no_std`. We need to check if it's possible to use anyhow there.

Deliverables:

1. Replace Result<> by anyhow Result<> under `standalone/phost`, and `standalone/pruntime/enclave/src` (master branch)
2. Test the full stack (phala-node, phost, pruntime) and make sure they can build, and sync to at least 1k blocks.

Prepare the dev env: https://wiki.phala.network/en-us/docs/developer/run-a-local-development-network/
  - Mostly up-to-date, but pruntime & phost have been moved to standalone

### PR DESCRIPTION
**Pull request by KiChjang**
Fixes #174.

All three core blockchain applications (phala-node, phost and pruntime) builds successfully, but while running phost, the following error message results:

```
Connected to substrate at: ws://localhost:9944
bridge() exited with result: Err(SubxtRpcError(Metadata(ModuleNotFound("Phala"))))
```

I don't understand how this is possible when I have `phala-node` running in the background, and I'm pretty sure it does have the Phala module.

### CODE REVIEW

**Comment by h4x3rotab** [comment]
Thanks for the contribution! I tried to pull the latest `master` branch and started the three components. I didn't get the error you encountered. Could you sync your branch to the latest master branch and try again?

However, I did find some suspicious problems. Recently we have renamed "PhalaModule" to "Phala", but the storage prefix was not changed (`trait Store for ... as PhalaModule`). I'm checking it locally.

----

Question: As we are using `anyhow`, and it supports any Err type implementing `std::error::Error`, could we remove the wrapping code like the one at `standalone/phost/src/error.rs`?

**Comment by KiChjang** [comment]
@h4x3rotab Branch is already up to date with latest master, so it's likely that your suspicion is correct.

I took a look at what the file contains, and I would say that all the `From` trait implementations to convert errors from other crates are not necessary anymore. I can remove them in this PR and see if it still compiles.

**Comment by KiChjang** [comment]
@h4x3rotab It looks like the `Error` type in `standalone/phost/src/error.rs` doesn't quite implement the `std::error::Error` trait, so I'll have to implement that in order to make it compile.

**Comment by KiChjang** [comment]
@h4x3rotab I tried changing the storage prefix to `Phala`, and running it gives me a slightly different error:

```
Connected to substrate at: ws://localhost:9944
bridge() exited with result: Err(Metadata error: Module Phala not found

Caused by:
    Module Phala not found)
```

**Comment by h4x3rotab** [comment]
I pulled your latest code, compile from scratch, and started these daemons and they worked well. Especially it crashed right after the launch. Could you make sure `phost` has connected to the correct node? It connects to localhost:9944 by default, and you need to make sure the port is running the latest `phala-node` (did you run a standard Substrate on 9944 instead?).

**Comment by h4x3rotab** [comment]
> @h4x3rotab Branch is already up to date with latest master, so it's likely that your suspicion is correct.
> 
> I took a look at what the file contains, and I would say that all the `From` trait implementations to convert errors from other crates are not necessary anymore. I can remove them in this PR and see if it still compiles.

Please push the new commit. I can also help you test it on my local workstation.

**Comment by KiChjang** [comment]
@h4x3rotab My new commit properly uses `anyhow::Result`. Previously it was actually still using `std::result::Result`, because `anyhow::Result` only contains 1 type argument, whereas `std::result::Result` accepts 2, so when `Result<A, B>` is used, it actually uses `std::result::Result` instead of the intended `anyhow::Result`.

One notable change I made was to serialize/deserialize anyhow errors in pRuntime, please take a look and see if it makes any sense.

File: standalone/pruntime/enclave/src/contracts/mod.rs [hunk 1]
```diff
@@ -95,6 +96,28 @@ pub mod serde_balance {
...
+        String::serialize(&s, serializer)
+    }
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<Error, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
+        Ok(Error::msg(s))
+    }
+}
```
Comment (KiChjang) [inline standalone/pruntime/enclave/src/contracts/mod.rs:119]: @h4x3rotab This is the new change that I made in order to get it working for `Response`s that return an `anyhow::Error`.

File: standalone/phost/src/error.rs [hunk 2]
```diff
@@ -17,38 +13,20 @@ pub enum Error {
...
-        Error::UriError(error)
-    }
-}
-
-impl From<serde_json::error::Error> for Error {
-    fn from(error: serde_json::error::Error) -> Error {
-        Error::SerdeError(error)
-    }
-}
+impl error::Error for Error {}
```
Comment (KiChjang) [inline standalone/phost/src/error.rs:32]: This impl is necessary because anyhow expects all errors to implement `std::error::Error`.

**Comment by KiChjang** [comment]
> 
> 
> I pulled your latest code, compile from scratch, and started these daemons and they worked well. Especially it crashed right after the launch. Could you make sure `phost` has connected to the correct node? It connects to localhost:9944 by default, and you need to make sure the port is running the latest `phala-node` (did you run a standard Substrate on 9944 instead?).
> 
> My commands:
> 
> ```shell
> ./target/release/phala-node --dev --ws-port 29944 --tmp
> ./bin  # under standalone/pruntime/bin
> ./target/release/phost --dev --substrate-ws-endpoint ws://localhost:29944 --pruntime-endpoint http://localhost:8000
> ```

I followed your instructions and now it runs successfully. Thanks!

File: standalone/phost/src/error.rs (hunk 2 above)
Comment (h4x3rotab) [reply standalone/phost/src/error.rs:32]: LGTM

File: standalone/pruntime/enclave/src/contracts/mod.rs (hunk 1 above)
Comment (h4x3rotab) [reply standalone/pruntime/enclave/src/contracts/mod.rs:119]: LGTM

**Comment by jasl** [comment]
LGTM

**Comment by h4x3rotab** [comment]
Great work! Thank you very much!

**Pull request merged**